### ISSUE TITLE
Panel text wrapping does not work with markdown starting with space

### ISSUE THREAD

**Issue by Jacob-Barhak** [pyviz/panel#271]
When using panel to show markdown text, there are several issues that depend on the text.
If there are spaces at the beginning of the line then the panel object size is ignored and the text does not wrap. 

Expected behavior is wrapping the text and in case the text should not wrap, it should be cropped to avoid exceeding panel size.

To recreate use this code:
```
import panel
import bokeh

BokehDocument = bokeh.document.Document()


Text1 = panel.panel("""### This is a title:
This is a very long sentence that uses many words to not really tell anything important, yet it is long so that that we can see if the panel object wraps the text properly and if the size of the panel defined in panel is kept.
""", width=200, height=120)

Text2 = panel.panel("""### This is a title:
	This is a very long sentence that uses many words to not really tell anything important, yet it is long so that that we can see if the panel object wraps the text properly and if the size of the panel defined in panel is kept.
""", width=200, height=120)


PanelDocument = panel.Column(Text1,Text2)
DocumentForOutput = PanelDocument._get_root(BokehDocument)

Html = bokeh.embed.file_html(DocumentForOutput, bokeh.resources.CDN, 'Testing 123')
OutFile = open('TestOutput.html','w')
OutFile.write(Html)
OutFile.close()

```

The output html file is attached within this zip file:
[TestOutput.zip](https://github.com/pyviz/panel/files/2930643/TestOutput.zip)

This is done with Python 2.7.14 |Anaconda custom (64-bit)|  and panel 0.4.0

**Comment by jsignell** [comment]
Markdown strips leading spaces by design so that part is intentional. The object size should be honored with wrapping if needed. That bug is captured in #270.

**Comment by Jacob-Barhak** [comment]
Before you close this issue, let us establish that we are on the same page. 

The issue with the spaces was not about spaces being stripped. Look at the code, when spaces lead the sentence, then wrapping is disabled and the panel size is therefore not honored . So there are two issues here:
1. leading spaces cause no wrapping of text
2. long text is not cropped at panel size

Yes, they are related since fixing one may affect the other.

You mention that one part is intentional. What exactly is intentional and where is it documented? 

Hopefully we are on the same page.

**Comment by jsignell** [comment]
I would argue that panel does exactly what github markdown does. Leading spaces get trimmed unless there are 4 (which is the case you have above) in which case they are treated as a code block. 

### This is a title:
This is a very long sentence that uses many words to not really tell anything important, yet it is long so that that we can see if the panel object wraps the text properly and if the size of the panel defined in panel is kept.

### This is a title: (with 4 leading spaces)
    This is a very long sentence that uses many words to not really tell anything important, yet it is long so that that we can see if the panel object wraps the text properly and if the size of the panel defined in panel is kept.

### This is a title: (with 3 leading spaces)
   This is a very long sentence that uses many words to not really tell anything important, yet it is long so that that we can see if the panel object wraps the text properly and if the size of the panel defined in panel is kept.

**Comment by jbednar** [comment]
Right; this behavior can be confusing, but it's the expected and designed behavior for markdown, so I don't think we can do anything about it.

**Comment by Jacob-Barhak** [comment]
Well James,

You should either clip the text, or raise a warning.

Yet your explanation makes sense. I was not aware that panel was thinking
my text is code. I was just trying to write a few paragraphs.

This brings another issue.

Markdown is designed for programmers. Yet if you want other types of users
for panel, you may want to consider on the long run using text with some
simple modifiers. Yet that is a different issue.

Thanks for explaining.

       Jacob

**Comment by jbednar** [comment]
> You should either clip the text, or raise a warning.

I don't think that's clear at all; the spaces are valid Markdown syntax, indicating an indented code block; and we shouldn't warn for valid syntax.  But there are actually two ways to indicate code blocks in Markdown, and so we could consider stripping spaces and telling people to use triple backticks instead. I personally do always use the backticks because they are more explicit, so I'd be ok with that approach.

@philippjfr, @jsignell, and @jlstevens, can you vote on this issue?  Is avoiding this confusing behavior (by stripping leading spaces) worth causing other confusion by not fully supporting Markdown?

**Comment by Jacob-Barhak** [comment]
So James,

You have a point in adding a strip spaces option - it will avoid this
issue.  Yet if you add it, you will get issues with indented bullets that
are allowed as well. So going down that route will create another issue.

I think you should see who your users will be. If you target non
programmers, you will have to invest much more effort to construct a good
experience .

For now, I am satisfied by your explanations. The discussion here seemed
necessary so others can follow why this was happening. You now clarified
the issue and addressed other concerns. I rest. I am good.

Jacob

**Comment by jbednar** [comment]
Panel is definitely a Python tool, and so we have to assume some level of Python programming ability.  What we are shooting for is someone who is comfortable writing a Jupyter notebook; if you can do that, you should be able to use Panel.  And Jupyter includes Markdown cells, hence assuming text is Markdown.  

For stripping spaces, I think we'd only strip the shared prefix of whitespace for all lines other than the first, which would avoid causing problems with indented bullets.
